webrender/src/device/gfx/buffer.rs

```diff
@@ -4,14 +4,84 @@
 
 use hal;
 use hal::Device as BackendDevice;
-use rendy_memory::{Block, Heaps, MemoryBlock, MemoryUsageValue, Write};
+use rendy_memory::{Block, Heaps, MappedRange, MemoryBlock, MemoryUsageValue, Write};
 use smallvec::SmallVec;
 
 use std::cell::Cell;
 use std::mem;
 
 pub const DOWNLOAD_BUFFER_SIZE: usize = 10 << 20; // 10MB
 
+pub(super) struct PMBuffer<B: hal::Backend> {
+    pub buffer: B::Buffer,
+    pub memory_block: MemoryBlock<B>,
+    pub coherent: bool,
+    pub height: u64,
+    pub size: u64,
+    pub state: Cell<hal::buffer::State>,
+    pub non_coherent_atom_size_mask: u64,
+    pub transit_range_end: u64,
+}
+
+impl<B: hal::Backend> PMBuffer<B> {
+    pub(super) fn map<'a>(&'a mut self, device: &B::Device, size: Option<u64>) -> (MappedRange<'a, B>, u64) {
+        let size = size.unwrap_or(self.size);
+        (self.memory_block.map(&device, 0..size).expect("Mapping memory block failed"), size)
+    }
+
+    pub(super) fn unmap(&mut self, device: &B::Device) {
+        self.memory_block.unmap(device);
+    }
+
+    pub(super) unsafe fn flush_mapped_ranges(
```

Comment:
do we need this now as we are using rendy-memory?

Reply:
I wanted to keep this, since rendy-memory's `MappedRange` has a lifetime, so we can't transfer it if we wan't write the mapped memory in the `RendererBackend` thread.
My idea would be store a pointer in the `RendererBackend` thread, which points to the start of the mapped memory region. We could use this ptr to write the GPUCache data in place and after that we send the ranges to the `Renderer` thread where we call  `flush_mapped_ranges` with those ranges.

Reply:
IIRC from our discussion on the call, we don't need to keep the mapped range if we use rendy-memory. We can rely on the fact that internally rendy-memory keeps the mapping open (since it has to). So what we need to do is whenever we need to write something, we ask rendy-memory for the mapped range, write it, and then let it go. There should be no need to keep it around and/or transfer anywhere.

Reply:
Removed this in the last commit: https://github.com/szeged/webrender/pull/297/commits/ed9e894391d366dfc69d34e7ef4b33250011e111#diff-5dbe2f57bb9273f9a363bcb71ec6834eL36